### ISSUE TITLE
How to reconstruct zeroes in my sparse vector?

### ISSUE THREAD

**Issue by nafizh** [fchollet/keras#4431]
I have 8000 size input sparse vectors in my keras autoencoder. For each vector, all the non-zero values sum to 1. But I am not being able to make the autoencoder produce the zeroes as zeroes rather some non-zero real values are being produced. These numbers are sometimes also in the range of my original non-zero values, hence I cannot use some kind of threshold to categorize them as zeroes. My network is with one hidden layer of 512 neurons. I am using mse as loss function and adagrad as the optimzer. I am also using relu as my activation functions.

```
from keras.layers import Input, Dense
from keras.models import Model
from keras.layers import GaussianNoise
# this is the size of our encoded representations
autoencoder = None
encoding_dim = 512  

# this is our input placeholder
input_img = Input(shape=(8000,))
encoded = GaussianNoise(1.0)(input_img)
encoded = Dropout(0.5)(input_img) 
# "encoded" is the encoded representation of the input
encoded = Dense(encoding_dim, activation='relu')(encoded)

# "decoded" is the lossy reconstruction of the input
encoded = Dropout(0.5)(encoded)
decoded = Dense(8000, activation = 'relu')(encoded)

# this model maps an input to its reconstruction
autoencoder = Model(input=input_img, output=decoded)

# this model maps an input to its encoded representation
encoder = Model(input=input_img, output=encoded)

# create a placeholder for an encoded (32-dimensional) input
encoded_input = Input(shape=(encoding_dim,))
# retrieve the last layer of the autoencoder model
decoder_layer = autoencoder.layers[-1]
# create the decoder model
decoder = Model(input=encoded_input, output=decoder_layer(encoded_input))

autoencoder.compile(optimizer='adagrad', loss='mse')

autoencoder.fit(x_train, x_train,
                nb_epoch= 200,
                batch_size=40,
                shuffle=True)
```

These are some values that should be zeroes-

```
[ 0.16505188  0.08558722  0.1612076   0.16082852  0.13480382  0.12213098
  0.02518164  0.18657388  0.19825849  0.16723372  0.06943481  0.17134313
  0.08403686  0.09241731  0.09426595  0.15179422  0.08171938  0.12384494
  0.          0.1315141   0.00397497  0.          0.          0.
  0.00821247  0.08011883  0.01093781  0.0126917   0.          0.04404544
  0.          0.          0.          0.          0.01234959  0.09994393
  0.          0.          0.          0.          0.11900149  0.          0.1048274
  0.20034248  0.1484586   0.13971123  0.0221523   0.20293164  0.10758945
  0.13483508]
```

And these are some values that are constructed rightly as non-zeroes(there are some zeroes though).
```
[ 0.2588819   0.10506743  0.          0.20462847  0.13885625  0.16941361
  0.23944859  0.12471424  0.09177548  0.16154648  0.07812363  0.20357776
  0.11772975  0.12835549  0.14947671  0.09359284  0.14273046  0.07017358
  0.02605285  0.          0.09658372  0.00264598  0.21479762  0.
  0.11528967  0.0556695   0.02868088  0.13447164  0.0609164   0.1478954
  0.17546555  0.04057987  0.16114648  0.17044659  0.16331613  0.11542476
  0.1547848   0.19915491  0.12712993  0.12984735  0.19595313  0.28850603
  0.16067727  0.03010978  0.30560112  0.03643312  0.          0.24497849
  0.03941336  0.17225127  0.14811273  0.15813884  0.09201842  0.05885779
  0.10273241  0.1184406   0.06972356  0.09816362  0.20082688  0.11600598
  0.12679152  0.18858148  0.15708457  0.11016279  0.08384509  0.
  0.15034403  0.05045035  0.17912017  0.07692797  0.10639004  0.03966355
  0.2560263   0.25950629  0.17312737  0.15898249  0.21245323  0.          0.]
```

Can anybody suggest me what am I doing wrong?

**Comment by nafizh** [comment]
I have closed this issue as it is not directly connected to keras. But any input would still be welcome.
